### DISCUSSION TITLE
C++ support / API improvements Proposal

### DISCUSSION THREAD

**Discussion by ManavGhaiCode** [godotengine/godot-proposals#7840] [Scripting]
## Introduction

Hi, I am a 13 year old developer named Manas for the who was using Unity.
So, I you might thing why I am opening this discussion as Godot have solid C# support.

Well it is not C#, I was very happy to know the Godot supported C++ as I used C++ for programing well before C# and wished that Unity also use C++

### What I don't like about Godot with C++

So, when I started to see how to use C++ in Godot, I was very confused, how it work why I can't use it directly and other stuff. I get it just a bit, but I think if you put a bunch of header files into a init file which start up the game and you include the Physics and Renderer in the header files, that will allow us to use C++ without any fuzz.

Tell me if I am wrong, but all the function which need to binded add stuff. Can be remove.
By just making one engine file which sets up the scenes and starting the game, which need to be done anyway and is what I think Godot does to handle GDscript file.

## Proposal

So, I think if the Godot team just put a little bit for their budget towards C++ support by the end of the dev cycle of 5.0.0 we can have a great API for C++ in Godot.

I know it will be hard to make a new translation layer/API for C++ will be very hard and time taking and no means cheap, but it will relay level up the fell of Godot to people like me who love low-level programing languages like C++

One more thing the API I have been taking about is just a translation of the API form GDscript to C++ but a little more like the one in Unity (put a lot of time seeing how the Physics work with RigidBody more the I would like to admit)

like for moving a top-down player:

```c++
#include<Physics>

class Player : public RigidBody2d {
    public:

        void Update override {
             base::Update();

              int x = Input.GetKeyDown(KEY_A) - Input.GetKeyDown(KEY_D);
              int y = Input.GetKeyDown(KEY_W) - Input.GetKeyDown(KEY_S);

              Velocity = new Vector2 (x, y);
        }
}
```

### Why I don't use GDscript

Now, to be up front I would say GDscript a language like Python, is not what you WANT WHEN MAKING A GAME.
Now, this is all up to if you like the syntax the snake but, I hate Python more the Unity's pricing system.

Now GDscript is good, way better than Python. And I appreciate all the work done by the Godot devs and fell very inspired by it. I think we will have a better time if Godot devs put more time in C++

### End

I will love to contribute to Godot and work on this, but sifting through 50k commit will think quite a bit of time.
But a soon as I can I will start to contribute to Godot.

I love the engine and how it work, it has it's down falls. Coming form a Unity background I would love to have a more GameObject based setup for example which is better in my opinion.

I will stop now, got to do home work than write this at the time.

Thank for reading it.

**Comment by AThousandShips** [comment]
Godot already has support for C++, [here](https://github.com/godotengine/godot-cpp)

**Reply by Zireael07** [reply]
They want something that doesn't go through GDExtension (a very similar issue was recently opened on godot-proposals)

**Reply by AThousandShips** [reply]
Oh that wasn't expressed so that wasn't obvious

**Reply by ManavGhaiCode** [reply]
Sorry for not being obvious, but I did state that I was very confused on how to get C++ working.
